### ISSUE TITLE
goproxyClient doesn't list version >=v2

### ISSUE THREAD

**Issue by mkmik** [kubernetes-sigs/cluster-api#7700]
**What steps did you take and what happened:**

```console
$ clusterctl upgrade apply --infrastructure capa-system/aws:v2.0.2 -v 10
No default config file available
Checking cert-manager version...
Cert-manager is already up to date
Performing upgrade...
Fetching File="metadata.yaml" Provider="cluster-api" Type="CoreProvider" Version="v1.3.0"
Fetching File="metadata.yaml" Provider="aws" Type="InfrastructureProvider" Version="v1.5.2"
Error: invalid target version: version v2.0.2 for the provider capa-system/infrastructure-aws does not match any release series
````

**What did you expect to happen:**

```
...
Fetching File="metadata.yaml" Provider="aws" Type="InfrastructureProvider" Version="v2.0.2"
...
```

**Anything else you would like to add:**

It worked when I disabled goproxy with `export GOPROXY=off`.

I suspect that `goproxyClient` used from `cmd/clusterctl/client/repository/repository_github.go` (presumably to avoid hitting the GH rate limiter) does work well when the version of the component is v2 (like the latest version of the aws provider)

**Environment:**

- Cluster-api version: v1.3.0
- minikube/kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
/area clusterctl
/area upgrades
/area artifacts
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by chrischdi** [comment]
Oh, this sounds bad 😞 👀

**Comment by 24sama** [comment]
If a project doesn't follow the go mod semver, maybe there will be no v2, v3 tag int he "proxy.golang.org".

**Comment by chrischdi** [comment]
> If a project doesn't follow the go mod semver, maybe there will be no v2, v3 tag in the "proxy.golang.org".

I can guess what you are thinking about, but could you explain a bit more or give an example?

**Comment by 24sama** [comment]
For example:
The latest version of [CAPKK](https://github.com/kubesphere/kubekey) is v3.0.2, but it's only a git tag, and the go mod file is still like this:
```
module github.com/kubesphere/kubekey
```
So, because of this, this link can not list any version:
https://proxy.golang.org/github.com/kubesphere/kubekey/v3/@v/list

And the https://proxy.golang.org/github.com/kubesphere/kubekey/@v/list
It only lists v0.* and v1.* versions.

**Comment by chrischdi** [comment]
Yes, this is correct. cc @fabriziopandini 

Curious about the reasons of doing that, but that's not a discussion for here 😊

Thanks for your inputs @mkmik and @24sama

@fabriziopandini this leads to the question what to do with it. I'll add it to the agenda of todays office hours :-)

**Comment by chrischdi** [comment]
For reference: the goproxy functionality was introduced in #7192 and is included in CAPI v1.3
